From 6d9e1d3bdeb8dee3d482cd98bcb598730c2fca7a Mon Sep 17 00:00:00 2001
From: Kyrylo Yatsenko <hedrok@gmail.com>
Date: Mon, 20 Apr 2026 23:02:27 +0300
Subject: [PATCH 1/2] pathd: add optional protocol to 'no mpls-te import'

Command `mpls-te import ospfv2|ospfv3|isis` had `no` version without
last option, so frr-reload.py failed when it was removed.

To fix:
* Merge `no` and not `no` versions
* Make protocol optional parameter in case of `no`
* If protocol is given in `no` version, check that currently import for
  this protocol is activated, otherwise ignore command.
* Create helper function `path_ted_start_importing_igp`.

Signed-off-by: Kyrylo Yatsenko <hedrok@gmail.com>
(cherry picked from commit d421a3360bdb6eba18777388325b003b8bc985bc)
---
 pathd/path_ted.c | 92 ++++++++++++++++++++++++++++--------------------
 1 file changed, 53 insertions(+), 39 deletions(-)

diff --git a/pathd/path_ted.c b/pathd/path_ted.c
index 05bd8cbf93..9dcd398d43 100644
--- a/pathd/path_ted.c
+++ b/pathd/path_ted.c
@@ -25,11 +25,12 @@
 static struct ls_ted *path_ted_create_ted(void);
 static void path_ted_register_vty(void);
 static void path_ted_unregister_vty(void);
-static uint32_t path_ted_start_importing_igp(const char *daemon_str);
+static uint32_t path_ted_start_importing_igp(enum igp_import daemon);
 static uint32_t path_ted_stop_importing_igp(void);
 static enum zclient_send_status path_ted_link_state_sync(void);
 static void path_ted_timer_handler_sync(struct event *event);
 static void path_ted_timer_handler_refresh(struct event *event);
+static enum igp_import str2igp_import(const char *str);
 
 extern struct zclient *pathd_zclient;
 
@@ -66,18 +67,18 @@ uint32_t path_ted_teardown(void)
  * @return		true if ok
  *
  */
-uint32_t path_ted_start_importing_igp(const char *daemon_str)
+uint32_t path_ted_start_importing_igp(enum igp_import daemon)
 {
 	uint32_t status = 0;
 
-	if (strcmp(daemon_str, "ospfv2") == 0)
-		ted_state_g.import = IMPORT_OSPFv2;
-	else if (strcmp(daemon_str, "ospfv3") == 0) {
-		ted_state_g.import = IMPORT_UNKNOWN;
-		return 1;
-	} else if (strcmp(daemon_str, "isis") == 0)
-		ted_state_g.import = IMPORT_ISIS;
-	else {
+	switch (daemon) {
+	case IMPORT_OSPFv2:
+	case IMPORT_ISIS:
+		ted_state_g.import = daemon;
+		break;
+	case IMPORT_OSPFv3:
+	case IMPORT_UNKNOWN:
+	default:
 		ted_state_g.import = IMPORT_UNKNOWN;
 		return 1;
 	}
@@ -406,7 +407,8 @@ DEFUN (no_path_ted,
 /* clang-format off */
 DEFPY(path_ted_import,
        path_ted_import_cmd,
-       "mpls-te import <ospfv2|ospfv3|isis>$import_daemon",
+       "[no] mpls-te import ![ospfv2|ospfv3|isis]$import_daemon",
+       NO_STR
        "Enable the TE database (TED) fill with remote igp data\n"
        "import\n"
        "Origin ospfv2\n"
@@ -414,34 +416,32 @@ DEFPY(path_ted_import,
        "Origin isis\n")
 /* clang-format on */
 {
-
-	if (ted_state_g.enabled)
-		if (path_ted_start_importing_igp(import_daemon)) {
-			vty_out(vty, "Unable to start importing\n");
-			return CMD_WARNING;
-		}
-	return CMD_SUCCESS;
-}
-
-/* clang-format off */
-DEFUN (no_path_ted_import,
-       no_path_ted_import_cmd,
-       "no mpls-te import",
-       NO_STR
-       NO_STR
-       "Disable the TE Database fill with remote igp data\n")
-/* clang-format on */
-{
-
-	if (ted_state_g.import) {
-		if (path_ted_stop_importing_igp()) {
-			vty_out(vty, "Unable to stop importing\n");
-			return CMD_WARNING;
+	enum igp_import daemon;
+
+	if (no) {
+		if (ted_state_g.import) {
+			/* If no import_daemon given - disable current, otherwise check */
+			if (import_daemon != NULL) {
+				daemon = str2igp_import(import_daemon);
+				if (daemon != ted_state_g.import) {
+					PATH_TED_DEBUG("%s: PATHD-TED: Requested to disable import for %s, but import is active for another daemon. Ignoring.",
+						       __func__, import_daemon);
+					return CMD_SUCCESS;
+				}
+			}
+			if (path_ted_stop_importing_igp()) {
+				vty_out(vty, "Unable to stop importing\n");
+				return CMD_WARNING;
+			}
 		} else {
-			PATH_TED_DEBUG(
-				"%s: PATHD-TED: Importing igp data already OFF",
-				__func__);
+			PATH_TED_DEBUG("%s: PATHD-TED: Importing igp data already OFF", __func__);
 		}
+	} else {
+		if (ted_state_g.enabled)
+			if (path_ted_start_importing_igp(str2igp_import(import_daemon))) {
+				vty_out(vty, "Unable to start importing\n");
+				return CMD_WARNING;
+			}
 	}
 	return CMD_SUCCESS;
 }
@@ -505,6 +505,22 @@ uint32_t path_ted_config_write(struct vty *vty)
 	return 0;
 }
 
+/**
+ * Convert string to igp_import
+ */
+static enum igp_import str2igp_import(const char *str)
+{
+	if (!str)
+		return IMPORT_UNKNOWN;
+	if (strcmp(str, "isis") == 0)
+		return IMPORT_ISIS;
+	else if (strcmp(str, "ospfv2") == 0)
+		return IMPORT_OSPFv2;
+	else if (strcmp(str, "ospfv3") == 0)
+		return IMPORT_OSPFv3;
+	return IMPORT_UNKNOWN;
+}
+
 /**
  * Register the fn's for CLI and hook for config show
  *
@@ -517,7 +533,6 @@ static void path_ted_register_vty(void)
 	install_element(SR_TRAFFIC_ENG_NODE, &path_ted_on_cmd);
 	install_element(SR_TRAFFIC_ENG_NODE, &no_path_ted_cmd);
 	install_element(SR_TRAFFIC_ENG_NODE, &path_ted_import_cmd);
-	install_element(SR_TRAFFIC_ENG_NODE, &no_path_ted_import_cmd);
 
 	install_element(CONFIG_NODE, &debug_path_ted_cmd);
 	install_element(ENABLE_NODE, &debug_path_ted_cmd);
@@ -537,7 +552,6 @@ static void path_ted_unregister_vty(void)
 	uninstall_element(SR_TRAFFIC_ENG_NODE, &path_ted_on_cmd);
 	uninstall_element(SR_TRAFFIC_ENG_NODE, &no_path_ted_cmd);
 	uninstall_element(SR_TRAFFIC_ENG_NODE, &path_ted_import_cmd);
-	uninstall_element(SR_TRAFFIC_ENG_NODE, &no_path_ted_import_cmd);
 }
 
 /**
-- 
2.51.2

